### ISSUE TITLE
Is there a way to weigh losses by class?

### ISSUE THREAD

**Issue by pfuerste** [open-mmlab/mmdetection#6759]
I am trying to weigh the contribution of predictions according to their class. For example, a detection of class a should not have as big a impact on learning as a detection of class b.

I looked into the losses and the weighted_loss decorator, but there is no class information in there and I did not find anything else. Is there a way to weigh the losses according to class in configs?

**Comment by taofuyu** [comment]
I wrote this code last month when training RetinaNet.

In config:
train_cfg=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.45,
            min_pos_iou=0,
            ignore_iof_thr=-1,
	    use_iou_supression=False),
        allowed_border=-1,
        pos_weight=[1, 2, 5, 1, 1, 3], #weight for each class
        neg_weight=1.0,
        .........

In anchor_head.py, function _get_targets_single():
#set different weight for different category
assert(self.num_classes==len(self.train_cfg.pos_weight))
for i in range(self.num_classes):
    category_pos_inds = pos_inds[sampling_result.pos_gt_labels==i]
    label_weights[category_pos_inds] = self.train_cfg.pos_weight[i]

The point is,  change the values of **label_weights**

**Comment by pfuerste** [comment]
Looks good, thanks! 
One question still: You are just changing the label_weights, not the bbox_weights.
My first thought on this would be that wrong classification of a found detected bounding box of a higher weighted class would be penalized more than that of any other class (given more than 2 classes).
Should I just alter the bbox_weights in the same way to get better bbox coordinates for certain classes?

**Comment by taofuyu** [comment]
> Looks good, thanks! One question still: You are just changing the label_weights, not the bbox_weights. My first thought on this would be that wrong classification of a found detected bounding box of a higher weighted class would be penalized more than that of any other class (given more than 2 classes). Should I just alter the bbox_weights in the same way to get better bbox coordinates for certain classes?

Yes，just follow "label_weights[category_pos_inds] = self.train_cfg.pos_weight[i]":
bbox_weights[category_pos_inds, :] = self.train_cfg.pos_bbox_weight[i]

To make things more clear, the whole _get_targets_single() in my code:
`def _get_targets_single(self,
                            flat_anchors,
                            valid_flags,
                            gt_bboxes,
                            gt_bboxes_ignore,
                            gt_labels,
                            img_meta,
                            label_channels=1,
                            unmap_outputs=True):
        """Compute regression and classification targets for anchors in a
        single image.

        Args:
            flat_anchors (Tensor): Multi-level anchors of the image, which are
                concatenated into a single tensor of shape (num_anchors ,4)
            valid_flags (Tensor): Multi level valid flags of the image,
                which are concatenated into a single tensor of
                    shape (num_anchors,).
            gt_bboxes (Tensor): Ground truth bboxes of the image,
                shape (num_gts, 4).
            img_meta (dict): Meta info of the image.
            gt_bboxes_ignore (Tensor): Ground truth bboxes to be
                ignored, shape (num_ignored_gts, 4).
            img_meta (dict): Meta info of the image.
            gt_labels (Tensor): Ground truth labels of each box,
                shape (num_gts,).
            label_channels (int): Channel of label.
            unmap_outputs (bool): Whether to map outputs back to the original
                set of anchors.

        Returns:
            tuple:
                labels_list (list[Tensor]): Labels of each level
                label_weights_list (list[Tensor]): Label weights of each level
                bbox_targets_list (list[Tensor]): BBox targets of each level
                bbox_weights_list (list[Tensor]): BBox weights of each level
                num_total_pos (int): Number of positive samples in all images
                num_total_neg (int): Number of negative samples in all images
        """
        inside_flags = anchor_inside_flags(flat_anchors, valid_flags,
                                           img_meta['img_shape'][:2],
                                           self.train_cfg.allowed_border)
        if not inside_flags.any():
            return (None, ) * 7
        # assign gt and sample anchors
        anchors = flat_anchors[inside_flags, :]

        assign_result = self.assigner.assign(
            anchors, gt_bboxes, gt_bboxes_ignore,
            None if self.sampling else gt_labels)
        sampling_result = self.sampler.sample(assign_result, anchors,
                                              gt_bboxes)

        num_valid_anchors = anchors.shape[0]
        bbox_targets = torch.zeros_like(anchors)
        bbox_weights = torch.zeros_like(anchors)
        labels = anchors.new_full((num_valid_anchors, ),
                                  self.num_classes,
                                  dtype=torch.long)
        label_weights = anchors.new_zeros(num_valid_anchors, dtype=torch.float)

        pos_inds = sampling_result.pos_inds
        neg_inds = sampling_result.neg_inds
        if len(pos_inds) > 0:
            if not self.reg_decoded_bbox:
                pos_bbox_targets = self.bbox_coder.encode(
                    sampling_result.pos_bboxes, sampling_result.pos_gt_bboxes)
            else:
                pos_bbox_targets = sampling_result.pos_gt_bboxes
            bbox_targets[pos_inds, :] = pos_bbox_targets
            
            if gt_labels is None:
                # Only rpn gives gt_labels as None
                # Foreground is the first class since v2.5.0
                labels[pos_inds] = 0
            else:
                labels[pos_inds] = gt_labels[
                    sampling_result.pos_assigned_gt_inds]
            
            #set different weight for different category
            assert(self.num_classes==len(self.train_cfg.pos_weight))
            assert(self.num_classes==len(self.train_cfg.pos_bbox_weight))
            for i in range(self.num_classes):
                category_pos_inds = pos_inds[sampling_result.pos_gt_labels==i]
                label_weights[category_pos_inds] = self.train_cfg.pos_weight[i]
                bbox_weights[category_pos_inds, :] = self.train_cfg.pos_bbox_weight[i]
        if len(neg_inds) > 0:
            label_weights[neg_inds] = 1.0

        # map up to original set of anchors
        if unmap_outputs:
            num_total_anchors = flat_anchors.size(0)
            labels = unmap(
                labels, num_total_anchors, inside_flags,
                fill=self.num_classes)  # fill bg label
            label_weights = unmap(label_weights, num_total_anchors,
                                  inside_flags)
            bbox_targets = unmap(bbox_targets, num_total_anchors, inside_flags)
            bbox_weights = unmap(bbox_weights, num_total_anchors, inside_flags)

        return (labels, label_weights, bbox_targets, bbox_weights, pos_inds,
                neg_inds, sampling_result)`

You can compare the difference between mm raw code.
